File: packages/sdk-python/rollgate/retry.py
"""
Retry utility with exponential backoff and jitter.
"""

import asyncio
import random
from dataclasses import dataclass, field
from typing import Callable, TypeVar, Optional, Awaitable, Generic

from rollgate.errors import RollgateError

T = TypeVar("T")


@dataclass
class RetryConfig:
    """Configuration for retry behavior."""

    max_retries: int = 3
    """Maximum number of retry attempts."""

    base_delay_ms: int = 100
    """Base delay in milliseconds."""

    max_delay_ms: int = 10000
    """Maximum delay in milliseconds."""

    jitter_factor: float = 0.1
    """Jitter factor 0-1 to randomize delays."""


@dataclass
class RetryResult(Generic[T]):
    """Result of a retry operation."""

    success: bool
    data: Optional[T] = None
    error: Optional[Exception] = None
    attempts: int = 1


DEFAULT_RETRY_CONFIG = RetryConfig()


def calculate_backoff(attempt: int, config: RetryConfig) -> float:
    """
    Calculate backoff delay with exponential increase and jitter.

    Args:
        attempt: Current attempt number (0-indexed)
        config: Retry configuration

    Returns:
        Delay in seconds
    """
    # Exponential: base_delay * 2^attempt
    exponential_delay = config.base_delay_ms * (2**attempt)

    # Cap at max_delay
    capped_delay = min(exponential_delay, config.max_delay_ms)

    # Add jitter: random value between -jitter and +jitter
    jitter = capped_delay * config.jitter_factor * (random.random() * 2 - 1)

    delay_ms = max(0, capped_delay + jitter)
    return delay_ms / 1000.0  # Convert to seconds


def is_retryable_error(error: Exception) -> bool:
    """
    Check if an error is retryable.

    Args:
        error: The exception to check

    Returns:
        True if the error should be retried
    """
    if isinstance(error, RollgateError):
        return error.retryable

    message = str(error).lower()

    # Network errors (always retry)
    network_indicators = [
        "econnrefused",
        "etimedout",
        "enotfound",
        "econnreset",
        "network",
        "connection",
        "timeout",
        "dns",
    ]
    if any(indicator in message for indicator in network_indicators):
        return True

    # HTTP 5xx errors (server issues, retry)
    server_errors = ["500", "502", "503", "504"]
    if any(code in message for code in server_errors):
        return True

    # Rate limiting (retry with backoff)
    if "429" in message or "too many requests" in message:
        return True

    # HTTP 4xx errors (client errors, don't retry)
    client_errors = ["400", "401", "403", "404"]
    if any(code in message for code in client_errors):
        return False

    return False


async def fetch_with_retry(
    fn: Callable[[], Awaitable[T]],
    config: Optional[RetryConfig] = None,
) -> RetryResult[T]:
    """
    Execute an async function with retry logic and exponential backoff.

    Args:
        fn: Async function to execute
        config: Retry configuration

    Returns:
        RetryResult with success status and data/error
    """
    cfg = config or DEFAULT_RETRY_CONFIG
    last_error: Optional[Exception] = None

    for attempt in range(cfg.max_retries + 1):
        try:
            data = await fn()
            return RetryResult(success=True, data=data, attempts=attempt + 1)
        except Exception as error:
            last_error = error

            # Don't retry non-retryable errors
            if not is_retryable_error(error):
                return RetryResult(success=False, error=error, attempts=attempt + 1)

            # Don't sleep after the last attempt
            if attempt < cfg.max_retries:
                delay = calculate_backoff(attempt, cfg)
                await asyncio.sleep(delay)

    return RetryResult(
        success=False,
        error=last_error or Exception("Retry exhausted"),
        attempts=cfg.max_retries + 1,
    )


async def retry_async(
    fn: Callable[[], Awaitable[T]],
    config: Optional[RetryConfig] = None,
) -> T:
    """
    Execute an async function with retry, raising on failure.

    Args:
        fn: Async function to execute
        config: Retry configuration

    Returns:
        Result of the function

    Raises:
        Exception: The last error if all retries fail
    """
    result = await fetch_with_retry(fn, config)

    if not result.success:
        raise result.error  # type: ignore

    return result.data  # type: ignore

File: packages/sdk-core/src/retry.ts
/**
 * Retry utility with exponential backoff and jitter
 */

export interface RetryConfig {
  /** Maximum number of retry attempts (default: 3) */
  maxRetries: number;
  /** Base delay in milliseconds (default: 100) */
  baseDelayMs: number;
  /** Maximum delay in milliseconds (default: 10000) */
  maxDelayMs: number;
  /** Jitter factor 0-1 to randomize delays (default: 0.1) */
  jitterFactor: number;
}

export const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 3,
  baseDelayMs: 100,
  maxDelayMs: 10000,
  jitterFactor: 0.1,
};

export interface RetryResult<T> {
  success: boolean;
  data?: T;
  error?: Error;
  attempts: number;
}

/**
 * Sleep for a given number of milliseconds
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Calculate backoff delay with exponential increase and jitter
 */
export function calculateBackoff(attempt: number, config: RetryConfig): number {
  // Exponential: baseDelay * 2^attempt
  const exponentialDelay = config.baseDelayMs * Math.pow(2, attempt);

  // Cap at maxDelay
  const cappedDelay = Math.min(exponentialDelay, config.maxDelayMs);

  // Add jitter: random value between -jitter and +jitter
  const jitter = cappedDelay * config.jitterFactor * (Math.random() * 2 - 1);

  return Math.max(0, cappedDelay + jitter);
}

/**
 * Check if an error is retryable
 */
export function isRetryableError(error: unknown): boolean {
  if (!(error instanceof Error)) {
    return false;
  }

  // Check if error has explicit retryable property (e.g., RollgateError)
  if ("retryable" in error && typeof error.retryable === "boolean") {
    return error.retryable;
  }

  const message = error.message.toLowerCase();

  // Network errors (always retry)
  if (message.includes("econnrefused")) return true;
  if (message.includes("etimedout")) return true;
  if (message.includes("enotfound")) return true;
  if (message.includes("econnreset")) return true;
  if (message.includes("network")) return true;
  if (message.includes("fetch failed")) return true;
  if (message.includes("aborted")) return true;

  // HTTP 5xx errors (server issues, retry)
  if (message.includes("500")) return true;
  if (message.includes("502")) return true; // Bad Gateway
  if (message.includes("503")) return true; // Service Unavailable
  if (message.includes("504")) return true; // Gateway Timeout

  // Rate limiting (retry with backoff)
  if (message.includes("429")) return true;
  if (message.includes("too many requests")) return true;

  // HTTP 4xx errors (client errors, don't retry)
  if (message.includes("400")) return false;
  if (message.includes("401")) return false;
  if (message.includes("403")) return false;
  if (message.includes("404")) return false;

  return false;
}

/**
 * Execute a function with retry logic and exponential backoff
 */
export async function fetchWithRetry<T>(
  fn: () => Promise<T>,
  config: Partial<RetryConfig> = {},
): Promise<RetryResult<T>> {
  const fullConfig: RetryConfig = { ...DEFAULT_RETRY_CONFIG, ...config };
  let lastError: Error | undefined;

  for (let attempt = 0; attempt <= fullConfig.maxRetries; attempt++) {
    try {
      const data = await fn();
      return {
        success: true,
        data,
        attempts: attempt + 1,
      };
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));

      // Don't retry non-retryable errors
      if (!isRetryableError(lastError)) {
        return {
          success: false,
          error: lastError,
          attempts: attempt + 1,
        };
      }

      // Don't sleep after the last attempt
      if (attempt < fullConfig.maxRetries) {
        const delay = calculateBackoff(attempt, fullConfig);
        await sleep(delay);
      }
    }
  }

  return {
    success: false,
    error: lastError || new Error("Retry exhausted"),
    attempts: fullConfig.maxRetries + 1,
  };
}
